### PR TITLE
F50x implementation

### PR DESCRIPTION
**Pull request by olliemath**
Closes https://github.com/charliermarsh/ruff/issues/905

This uses a vendored copy of `cformat.rs` from rustpython-vm to parse the format string. Then we perform our checks in a similar manner to the F52x checks.

### Extensions?

I wrote a version of this code that looked inside splatted expressions like `{"a": 1, **b, "c": 2}` (which the interpreter parses identically to `{"a": 1, **b, **{"c": 2}}`. It could detect if e.g. `"c"` from the earlier example was unused in a format string, however it seemed overly complicated for such a small edge-case, which is not handled in pyflakes.

Along with `CFormatString`, rustpython-vm provides a `CFormatBytes` that could easily be used to perform the same checks on `b"..." % ...` format expressions. On the other hand, I don't think this is super valuable: pyflakes doesn't provide it and without a lot of thought the error messages may look horrible (raw bytes).

### CODE REVIEW

**Comment by charliermarsh** [comment]
This looks _great_! Anything else you want to do here or good to merge from your perspective?

File: src/pyflakes/checks.rs [hunk 1]
```diff
@@ -23,6 +24,210 @@ fn has_star_args(args: &[Expr]) -> bool {
...
+                    return None; // Dynamic keys present
+                }
+            }
+        }
+
+        let missing: Vec<String> = summary
+            .keywords
+            .iter()
+            .filter(|k| !keywords.contains(k))
+            .cloned()
```
Comment (charliermarsh) [inline src/pyflakes/checks.rs:148]: We might be able to delay these `cloned` calls until the `else` block? Like make this `Vec<&str>` and then call `cloned` when creating (e.g.) `CheckKind::PercentFormatMissingArgument`? Really minor though (and didn't try it myself).

**Comment by olliemath** [comment]
> This looks _great_! Anything else you want to do here or good to merge from your perspective?

Right now the code picks up some issues pyflakes doesn't - I think the particular cases are useful, unless you'd prefer to have direct parity?

The only one I'm unsure about is in F505, where we are able to detect this while pyflakes is not:
```python
"%(a)d" % {"b": 1, **d}  # b is unused in the comprehension - flagged
```

For the average user of ruff, I think the fact that the above is detected but the following is _not detected_ violates the principal of least surprise:
```python
"%(a)d" % {**d, "b": 1}  # b is unused in the comprehension - not flagged
```
Because of the way this is parsed, we would have to check the keys 1 level down in any splatted dict literals to spot this.  What do you think of adding this extra check, or else giving up on splatted dicts entirely to keep the linting behaviour predictable?

File: src/pyflakes/checks.rs (hunk 1 above)
Comment (olliemath) [reply src/pyflakes/checks.rs:148]: Yeah, that's possible - we actually get a `Vec<&String>` and then put `missing.iter().map(|&s| s.clone()).collect()` into the check. 

I had assumed (possibly naively) that `cloned` would have zero overhead if the filter produces an empty iterator (which is what the if/else block is checking)?

Comment (olliemath) [reply src/pyflakes/checks.rs:148]: Unless the move is more about signalling intent than performance?

Comment (charliermarsh) [reply src/pyflakes/checks.rs:148]: Yeah I'm sure the difference is negligible if existent at all given that we're only cloning the empty `vec` anyway in that case. It's more about signaling intent, locally and broadly ("clone as late as possible"), but honestly don't feel strongly about changing it.

**Comment by charliermarsh** [comment]
Yeah I'd probably vote to ignore splatted dict literals given your reasoning above.

File: src/pyflakes/checks.rs (hunk 1 above)
Comment (olliemath) [reply src/pyflakes/checks.rs]: Makes sense - I've moved cloning later in the 3 functions where it was clear how to do so

**Comment by olliemath** [comment]
Cool, so I've disabled the introspection of splatted dicts, which gives behaviour more consistent (both internally and with pyflakes).

**Comment by olliemath** [comment]
I'd say this is good to go if the pipeline passes

**Comment by charliermarsh** [comment]
Awesome, will merge tonight.

**Pull request merged**